framework/scene_graph/components/image.cpp

```diff
@@ -1,370 +1,396 @@
+		default:
+			return fmt;
+	}
+}
+
+Image::Image(const std::string &name, std::vector<uint8_t> &&d, std::vector<Mipmap> &&m) :
+    Component{name},
+    data{std::move(d)},
+    format{VK_FORMAT_R8G8B8A8_UNORM},
+    mipmaps{std::move(m)}
+{
+}
+
+std::type_index Image::get_type()
+{
+	return typeid(Image);
+}
+
+const std::vector<uint8_t> &Image::get_data() const
+{
+	return data;
+}
+
+void Image::clear_data()
+{
+	data.clear();
+	data.shrink_to_fit();
+}
+
+VkFormat Image::get_format() const
+{
+	return format;
+}
+
+const VkExtent3D &Image::get_extent() const
+{
+	assert(!mipmaps.empty());
+	return mipmaps[0].extent;
+}
+
+const uint32_t Image::get_layers() const
+{
+	return layers;
+}
+
+const std::vector<Mipmap> &Image::get_mipmaps() const
+{
+	return mipmaps;
+}
+
+const std::vector<std::vector<VkDeviceSize>> &Image::get_offsets() const
+{
+	return offsets;
+}
+
+void Image::create_vk_image(Device const &device, VkImageViewType image_view_type, VkImageCreateFlags flags)
+{
+	assert(!vk_image && !vk_image_view && "Vulkan image already constructed");
+
+	vk_image = std::make_unique<core::Image>(device,
+	                                         get_extent(),
+	                                         format,
+	                                         VK_IMAGE_USAGE_SAMPLED_BIT | VK_IMAGE_USAGE_TRANSFER_DST_BIT,
+	                                         VMA_MEMORY_USAGE_GPU_ONLY,
+	                                         VK_SAMPLE_COUNT_1_BIT,
+	                                         to_u32(mipmaps.size()),
+	                                         layers,
+	                                         VK_IMAGE_TILING_OPTIMAL,
+	                                         flags);
+	vk_image->set_debug_name(get_name());
+
+	vk_image_view = std::make_unique<core::ImageView>(*vk_image, image_view_type);
+	vk_image_view->set_debug_name("View on " + get_name());
+}
+
+const core::Image &Image::get_vk_image() const
+{
+	assert(vk_image && "Vulkan image was not created");
+	return *vk_image;
+}
+
+const core::ImageView &Image::get_vk_image_view() const
+{
+	assert(vk_image_view && "Vulkan image view was not created");
+	return *vk_image_view;
+}
+
+Mipmap &Image::get_mipmap(const size_t index)
+{
+	assert(index < mipmaps.size());
+	return mipmaps[index];
+}
+
+uint32_t get_required_mipmaps_size(const VkExtent3D &extent)
+{
+	constexpr uint32_t channels    = 4;
+	auto               next_width  = std::max<uint32_t>(1u, extent.width);
+	auto               next_height = std::max<uint32_t>(1u, extent.height);
+	auto               next_size   = next_width * next_height * channels;
+	uint32_t           result      = 0;
+	while (true)
+	{
+		result += next_size;
+		// Next mipmap values
+		next_width  = std::max<uint32_t>(1u, next_width / 2);
+		next_height = std::max<uint32_t>(1u, next_height / 2);
+		next_size   = next_width * next_height * channels;
+		if (next_width == 1 && next_height == 1)
+		{
+			break;
+		}
+	}
+
+	return result;
+}

```

Comment:
You're missing the 1x1 level.
Maybe better:
```
	constexpr uint32_t channels = 4;
	auto               width    = std::max<uint32_t>(1, extent.width);
	auto               height   = std::max<uint32_t>(1, extent.height);
	auto               size     = width * height * channels;
	auto               result   = size;
	while (size != channels)
	{
		width  = (width + 1) / 2;
		height = (height + 1) / 2;
		size   = width * height * channels;
		result += size;
	}

	return result;
```

Reply:
Good point, but th suggested replacement fails as well for the degenerate case of a 1x1 image, so the while should be `do {...} while` I think.

Reply:
Ahm... no, I don't think it would fail.
result is initialized with the top-level size. If that top-level size is 1x1, it's correct.

Reply:
Sorry, you're right, I missed that.